### PR TITLE
docs: Upgrade & Rollback Testing

### PR DESCRIPTION
**Pull request by demirkayaender**
**What changed?**

- Added the Upgrade & Rollback Testing tech-review page.
- Clarified released schema tooling, adjacent-version rollback expectations, and the absence of a public end-to-end upgrade/downgrade test.
- Included small documentation consistency fixes identified during review of #436.

**Why?**

The CNCF tech review asks how Cadence validates upgrade, rollback, and second-upgrade paths. This change documents the public project coverage, clearly identifies the complete end-to-end sequence as not publicly tested, and gives operators a staging rehearsal for their exact backends and versions.

**How did you verify it?**

- Ran `yarn build` successfully.
- Checked the documented schema behavior and tooling against the public Cadence server source and release history.
- Reviewed changed files for broken links and style violations.

Personal GitHub Pages preview (required for site changes):
https://demirkayaender.github.io/Cadence-Docs/docs/tech-review/day-1-installation/rollout-upgrade-rollback/upgrade-rollback-testing

**Potential risks**

Documentation-only. The main risk is overstating compatibility or public test coverage; the page now conditions rollback support on release notes and marks the complete public end-to-end test as unavailable.

**Related changes**

- #436

### CODE REVIEW

File: blog/2026-07-07-batch-actions-ui/2026-07-07-batch-actions-ui.md [hunk 1]
```diff
@@ -94,6 +94,6 @@ Batch Actions close the gap between observing a problem and fixing it at scale.
 
 Upgrade to the latest **Cadence Web** release and give it a try. We'd love to hear what workflows you're managing at scale.
 
-- **Ask a question** — [#cadence on CNCF Slack](https://inviter.co/cncf)
+- **Ask a question** - [#cadence-users on CNCF Slack](https://inviter.co/cncf)
 - **Open an issue** — [cadence-web on GitHub](https://github.com/cadence-workflow/cadence-web/issues)
 - **Join the discussion** — [GitHub Discussions](https://github.com/cadence-workflow/cadence/discussions)
```
Comment (vladimir-gavrilenko) [inline blog/2026-07-07-batch-actions-ui/2026-07-07-batch-actions-ui.md:99]: nit: the rest of the list uses em-dashes, so it's better to either change all or none.

File: docs/07-operation-guide/02-maintain.md [hunk 2]
```diff
@@ -110,8 +110,10 @@ NOTE: Do not use “auto-setup” images to upgrade your schema. It's mainly for
 For how to apply database schema, refer to this doc: [SQL tool README](https://github.com/cadence-workflow/cadence/tree/master/tools/sql)
 [Cassandra tool README](https://github.com/cadence-workflow/cadence/tree/master/tools/cassandra)
 
-The tool makes use of a table called “schema_versions” to keep track of upgrading History. But there is no transaction guarantee for cross table operations. So in case of some error, you may need to fix or apply schema change manually.
-Also, the schema tool by default will upgrade schema to the latest, so no manual is required. ( you can also specify to let it upgrade to any place, like 0.14).
+Use the schema tools and files shipped with the server release you are deploying. Releases that include `cadence-server update-schema` can read the server's static configuration and update all configured persistence stores. For earlier releases, run `cadence-sql-tool` or `cadence-cassandra-tool` for each default and visibility store.
```
Comment (vladimir-gavrilenko) [inline docs/07-operation-guide/02-maintain.md]: what is earlier release here?

File: docs/11-tech-review/02-day-1-installation/03-rollout-upgrade-rollback/05-upgrade-rollback-testing.md [hunk 3]
```diff
@@ -8,3 +8,66 @@ keywords:
...
+
+On startup, Cadence compares the installed schema with the version required by the binary. Persistence must be at least as new as the binary expects. This asymmetric check is intentional: after an additive schema migration, the previous binary can run against the newer schema during rollback. See the [schema compatibility check](https://github.com/cadence-workflow/cadence/blob/master/tools/common/schema/handler.go) and [schema verification implementation](https://github.com/cadence-workflow/cadence/blob/master/common/persistence/schema/verify.go).
+
+Do not assume every schema or data migration is backward compatible. Check the target release notes before upgrading and do not roll a binary back across a release that documents an incompatible migration. Also:
+
+- Upgrade one minor release at a time. Rollback is supported between N and N+1. It is not maintained between N-1 and N+1. If you need to roll back past N, ask in [CNCF Slack `#cadence-users`](https://inviter.co/cncf) or ask the maintainers.
```
Comment (vladimir-gavrilenko) [inline docs/11-tech-review/02-day-1-installation/03-rollout-upgrade-rollback/05-upgrade-rollback-testing.md]: This repeats what is already said on the line 49. Would it make sense to compact the whole section a bit?

File: docs/11-tech-review/02-day-1-installation/03-rollout-upgrade-rollback/05-upgrade-rollback-testing.md [hunk 4]
```diff
@@ -8,3 +8,66 @@ keywords:
...
+
+A persistence backup is optional. Operators who want that extra precaution can take datastore and visibility backups with their provider's usual procedure. A restore is the recovery path if a schema change needs to be undone.
+
+## Schema compatibility during rollback
+
+Schema version numbers are independent of Cadence server release numbers. The schema tools record the current version in persistence and apply migrations in order. They do not provide down migrations. Before a schema ships in an official release, the upgrade is already part of that high-scale pre-release soak. Rollback is expected to keep the newer schema and restore the previous binary.
+
+The project aims to keep adjacent minor versions N and N+1 backward compatible, so rolling the server binary back from N+1 to N is the supported path unless the release notes document an incompatible migration. Compatibility between N-1 and N+1 is not maintained. If a farther rollback is necessary, ask in [CNCF Slack `#cadence-users`](https://inviter.co/cncf) or ask the maintainers before you attempt it.
```
Comment (vladimir-gavrilenko) [inline docs/11-tech-review/02-day-1-installation/03-rollout-upgrade-rollback/05-upgrade-rollback-testing.md]:
```suggestion
The project aims to keep adjacent minor versions N and N+1 backward compatible, so rolling the server binary back from N+1 to N is the supported path unless the release notes document an incompatible migration. Compatibility between N-1 and N+1 is not maintained. If a farther rollback is necessary, ask the maintainers in [`#cadence-users` on CNCF Slack](https://inviter.co/cncf) before you attempt it.
```

File: docs/11-tech-review/02-day-1-installation/03-rollout-upgrade-rollback/05-upgrade-rollback-testing.md [hunk 5]
```diff
@@ -8,3 +8,66 @@ keywords:
   - cadence downgrade testing
 ---
 
+Cadence upgrades persistence schemas before server processes. Schema migrations move forward, while a rollback normally restores the previous server binary without reverting the schema. This sequence lets old and new server processes share a cluster during a rolling deployment when the release's schema changes are backward compatible.
+
+Cadence validates the individual parts of this compatibility model in continuous integration. The project does not currently run a public end-to-end test that deploys two released server versions and performs the complete upgrade, downgrade, and second upgrade automatically. Operators should therefore rehearse the sequence below in a staging cluster before each production minor-version upgrade.
```
Comment (vladimir-gavrilenko) [inline docs/11-tech-review/02-day-1-installation/03-rollout-upgrade-rollback/05-upgrade-rollback-testing.md]:
```suggestion
Cadence tests these compatibility properties individually in continuous integration. The project does not run a public end-to-end upgrade/downgrade test across two released server versions.
```
optional: this is more concise, doesn't use the word "model" for just two tests, and shortens a long negative sentence that describes what a hypothetical test would do.

File: docs/07-operation-guide/02-maintain.md [hunk 6]
```diff
@@ -110,8 +110,10 @@ NOTE: Do not use “auto-setup” images to upgrade your schema. It's mainly for
...
+
+The tools use the `schema_version` and `schema_update_history` tables to track schema upgrades. There is no transaction guarantee across all DDL statements and version-table updates, so a partial failure can require manual repair before retrying.
+By default, the schema tool upgrades to the latest schema included with the checked-out release. You can also specify a target schema version.
```
Comment (vladimir-gavrilenko) [inline docs/07-operation-guide/02-maintain.md:117]:
```suggestion
The tools use the `schema_version` and `schema_update_history` tables to track schema upgrades. There is no transaction guarantee across all DDL statements and version-table updates, so a partial failure can require manual repair before retrying.

By default, the schema tool upgrades to the latest schema included with the checked-out release. You can also specify a target schema version.
```
nit: these two seems to be different thoughts?

**Review by vladimir-gavrilenko — APPROVED** [review]

File: docs/11-tech-review/02-day-1-installation/03-rollout-upgrade-rollback/05-upgrade-rollback-testing.md [hunk 7]
```diff
@@ -8,3 +8,66 @@ keywords:
...
+
+## What the project tests
+
+| Area | Project validation |
+| --- | --- |
+| Persistence backends | Server integration suites run against Cassandra, MySQL, PostgreSQL, and SQLite. Additional jobs exercise Elasticsearch, OpenSearch, Pinot, Kafka, and cross-cluster replication configurations. |
+| Schema updates | Unit and integration tests apply versioned migrations, exercise fresh-store setup, and verify the schema recorded in each configured default and visibility store. |
+| High-scale pre-release soak | Before a version is published as a stable GitHub Release, it is run at high scale, including schema upgrades against billions of workflow executions. Those builds are published as documented pre-release tags, for example [`v1.2.19-prerelease07`](https://github.com/cadence-workflow/cadence/releases/tag/v1.2.19-prerelease07). |
```
Comment (Bueller87) [inline docs/11-tech-review/02-day-1-installation/03-rollout-upgrade-rollback/05-upgrade-rollback-testing.md]: I believe this is true. But, can we backup the claim of 'billions'?  

Suggestion: Do we have a public test report or release document that supports the schema-upgrade claim? If not, it may be safer to narrow this to the testing we can verify publicly.

File: docs/11-tech-review/02-day-1-installation/03-rollout-upgrade-rollback/05-upgrade-rollback-testing.md (hunk 5 above)
Comment (Bueller87) [inline docs/11-tech-review/02-day-1-installation/03-rollout-upgrade-rollback/05-upgrade-rollback-testing.md]: **Lines**: 13-41

**Issue**: Are we answering how Cadence tested the complete upgrade, downgrade, and second upgrade path? This says no public end-to-end test exists and then documents a rehearsal for operators to perform.

**Suggestion**: Would it make sense to link an actual project test with versions, backends, checks, and results? If no such test exists, perhaps this part should be marked N/A with a short explanation.

File: docs/11-tech-review/02-day-1-installation/03-rollout-upgrade-rollback/05-upgrade-rollback-testing.md (hunk 7 above)
Comment (demirkayaender) [reply docs/11-tech-review/02-day-1-installation/03-rollout-upgrade-rollback/05-upgrade-rollback-testing.md]: Yes, we even presented about this in Open Source Summit in 2023. Back then monthly workflow execution count was 12B+.

File: docs/11-tech-review/02-day-1-installation/03-rollout-upgrade-rollback/05-upgrade-rollback-testing.md [hunk 8]
```diff
@@ -8,3 +8,66 @@ keywords:
...
+2. **Apply the N+1 schemas:** Use the schema tooling and schema files shipped with N+1. Apply the default-store and visibility-store changes before starting N+1 server processes. Use `cadence-cassandra-tool` or `cadence-sql-tool` for releases that provide the database-specific tools. Releases that include the configuration-driven `cadence-server update-schema` command can update the stores declared in the server configuration.
+3. **Roll forward to N+1:** Replace Cadence server processes gradually so N and N+1 coexist during the rollout. Wait for service membership and History shard ownership to stabilize between batches.
+4. **Validate N+1:** Confirm that workflows started on N continue, workers keep polling, new workflows start and complete, signals and queries succeed, timers fire, visibility queries return expected results, and service error and persistence latency metrics remain within the deployment's normal range.
+5. **Roll binaries back to N:** Restore the N server image without downgrading the schema. Repeat the workload and operational checks. A release note that marks a change as incompatible overrides the general rollback expectation.
```
[the remark this answers is not in the document]
Comment (demirkayaender) [reply docs/11-tech-review/02-day-1-installation/03-rollout-upgrade-rollback/05-upgrade-rollback-testing.md:40]: Schema upgrades are different than binary updates. Upgraded schemas cannot be rolled back like binaries.
